### DISCUSSION TITLE
What is Object Oriented Programming? Explain with an example.

### DISCUSSION THREAD

**Discussion by mishelshaji** [mishelshaji/Learn-DotNet#11] [Q&A] [answered]
Write what is oop, give appropriate examples, compare with functional programming

**Comment by sreelakshmiChandran** [comment]
# Object Oriented Programming(OOP)

OOP is programming that organizes software design around objects or data, rather than functions and logic. It relies on the concept of class and objects.

- It provides a clear structure for the programs.
- Faster and easier to execute.
- OOP makes it possible to create fully reusable applications with less code and shorter development time.
- There are many object-oriented programming languages, such as JavaScript, Python, C++, and Java.

## Structure 

- Class: Act as the blueprint for individual objects, attributes, and methods.
- Objects: Description of a class created with specifically defined data.
- Methods: Functions that are defined inside a class that describe the behaviors of an object. 
- Attributes:  Represent the state of an object.

Let's define a class by the name Student. Attributes of this class defines the properties of the Student such as name, age. Class is a collection of objects with the same properties.
```csharp
class Student
{
     public string Name;
     public  int Age;
}
```
Functions can be defined in the class which is called methods of the class.
```csharp
class student
{
     public string Name;
     public int Age;

void Display()
   {
       Console.WriteLine($"The name of the student is {Name}.")
   }
}
```
Here the Display() is a function that is used to display the name of the student.

## Principles of OOP

- Abstraction
- Encapsulation
- Inheritance
- Polymorphism

## Why OOP rather than Functional programming
Functional programming is based on mathematical functions. Its primary focus is on 'what to solve'.

- One of the major differences between the two approaches is that Functional programming focuses on data and their transformations of the functions whereas Object-oriented is based on objects and models.
- Fundamental elements of the Functional programming are variables and functions, and Object-oriented are objects and methods.
- Oop follows the imperative programming model, as well as functional programming, which follows the declarative programming model.
- Importance is not given to data but to functions in functional programming. Importance is given to data rather than procedures in oop.

**Comment by Nandana-04** [comment]
# Object Oriented Programming (OOP)

Object Oriented Programming (OOP) is a programming model where programs are organized around objects and data rather than action and logic . 

- OOP is faster and easier to execute
- It  provides a clear structure for the programs
- It  helps to keep the C# code DRY "Don't Repeat Yourself", and makes the code easier to maintain, modify and debug
- OOP makes it possible to create full reusable applications with less code and shorter development time

Structure
Class - It is a blueprint of an object that contains variables for storing data and functions to perform operations on the data .
```csharp
// Create a class named  "Student" with a variable Id .
class Student
{
   int Id = 123;
}
```
Object -  The software is divided into a number of small units called objects. The data and functions are built around these objects . It is created from a class .
```csharp
class Student
{
   int Id = 123;
}
// Create an object called "stud" and use it to print the value of Id .
Student stud = new Student();
Console.WriteLine(stud.Id);
```

The four basic principles of object-oriented programming are :

- Encapsulation
- Inheritance
- Polymorphism
- Abstraction

Difference between Functional Programming(FP) and Object Oriented Programming(OOP)

- Functional programming  paradigm emphasizes on the use of functions where each function performs a specific task but ,  Object 
oriented programming paradigm is based on object oriented concept  and classes are used where instance of objects are created .

- In FP , fundamental elements used are variables and functions and the data in the functions are immutable . OOP fundamental elements used are objects and methods and the data used here are mutable data.

- In FP , it doesn't have any  access specifier but , OOP has three access specifiers namely, Public, Private and Protected.

**Comment by ARAVIND0407** [comment]
# Object Oriented Programing [OOPs]
Object-oriented programming is a programming model that organizes software design around data, or objects, rather than functions and logic. An object can be defined as a data field that has unique attributes and behavior.

OOP focuses on the objects that developers want to manipulate rather than the logic required to manipulate them. This approach to programming is well-suited for programs that are large, complex and actively updated or maintained. This includes programs for manufacturing and design, as well as mobile applications.

The organization of an object-oriented program also makes the method beneficial to collaborative development, where projects are divided into groups. Additional benefits of OOP include code reusability, scalability and efficiency.

**Reply by mishelshaji** [reply]
Please include an example.

**Comment by akhilamp123** [comment]
**Object Oriented Programming (OOPs)**

Object oriented programming(OOPs) is a type of programming  that is based on objects rather than just functions and procedures. 
Individual objects are grouped into classes. OOPs implement real-world entities like inheritance, polymorphism, hiding etc into programming. It also allows binding data and code together. Classes and objects are the two main aspects of object-oriented programming.

**Class**

A class is an entity that determines how an object will behave and what the object will contain. To create a class ,we can use the class keyword:

````Csharp
class Car
{
	string color = "red";
}
````
When a variable is declared directly in a class, it is often referred to as a field (or attribute).

**Object**
An object can be defined as an instance of a class, and there can be multiple instances of a class in a program.

```csharp
class Car
    {
         public string color = "Blue";
    }

using OOPs;

Car obj = new Car();
Console.WriteLine(obj.color);
````
**Principles of OOPs**

- Inheritance
- Encapsulation
- Polymorphism
- Data Abstraction

**Inheritance**
Inheritance is a mechanism where the sub or child class inherits the properties and characteristics of super class.

**Encapsulation**

Encapsulation is an attribute of an object, and it contains all data which is hidden.

**Polymorphism**

Polymorphism is nothing but assigning behaviour or value in a subclass to something that was already declared in the main class.
Simply, polymorphism takes more than one form.

**Abstraction**

Abstraction is the concept of oop that shows only essential attributes and hides unnecessary information.

**OOps Vs Functional Programming**

**OOPs**

- This programming paradigm is based on object oriented concept. Classes are used where instance of objects are created.
- Fundamental elements used are objects and methods and the data used here are mutable data.
- Importance is given to data rather than procedures.
- It uses loops for iteration.
- It does not support parallel programming.


**Functional Programming**

- This programming paradigm emphasizes on the use of functions where each function performs a specific task.
- Fundamental elements used are variables and functions. The data in the functions are immutable(cannot be changed after creation).
- Importance is not given to data but to functions.
- It uses recursion for iteration.
- It is parallel programming supported.

**Comment by riya45810** [comment]
### Object-Oriented Programming in C#
Object-oriented programming is about creating objects that contain both data and methods, rather than writing procedures/methods that perform operations on the data(Procedure-oriented programming).
**Advantages:**

- Faster and easier to execute.
- Provides a clear structure for the programs.
-  Helps to keep the C# code "Don't repeat yourself" ,and makes the code easier to maintain, modify and debug.
- Makes it possible to create full reusable applications with less code and shorter development time.

Classes and objects are the two main aspects of OOP.
**class** is a template for objects, and a **object** is an instance of a class. When an individual object are created, they inherit all the variables and methods from the class.
```csharp
 // class created using class keyword
class Fruit 
{
    string name = "Apple";
    static void Main(string[] args){
    // object of the particular class created
    Fruit obj = new Fruit();   
    Console.WriteLine(obj.name);
}
}
```
**Principles of OOPs**

- Abstraction 
- Encapsulation
- Inheritance
- Polymorphism

**Abstraction**
It is the process of hiding certain unwanted details and only shows essential information to the user.it can be done with either abstract classes or interfaces. the abstract keyword is used for classes and methods.
**Encapsulation**
This binds the member function and data member into a single class. this also allows for abstraction.
**Inheritance**
Inheriting fields and methods from one class to another. the class that inherits from another class is called **Derived class**(child) and the class being inherited from is called as **Base class**(parent).
**Polymorphism**
It means "many forms", and it occurs when we have many classes that are related to each other by inheritance. they are two types: **Compile Time** polymorphism and **Runtime** polymorphism.

**Comment by amalapaul** [comment]
## OBJECT ORIENTED PROGRAMMING

Object-oriented programming (OOP) is a programming paradigm that uses objects and their interactions to design applications and computer programs.

- #### Class:

It is a collection of objects.

- ####  Objects:

Objects are basic building blocks of a C# OOP program. An object is a combination of data and methods. The data and the methods are called members of an object. In an OOP program, we create objects. These objects communicate together through methods. Each object can receive messages, send messages and process data.

Class is composed of three things name, attributes, and operations

```csharp
public class student
{
}
student objstudent=new student ();
```

From above we can say that Student object, named objstudent, has created out of the student class.

- ####  C# methods:

Methods are functions defined inside the body of a class. They are used to perform operations with the attributes of our objects. Methods bring modularity to our programs.


**There are some basic programming concepts in OOP:**

### a) C# inheritance
The inheritance is a way to form new classes using classes that have already been defined. The newly formed classes are called derived classes, the classes that we derive from are called base classes. Important benefits of inheritance are code reuse and reduction of complexity of a program. The derived classes (descendants) override or extend the functionality of the base classes (ancestors).

```csharp
new Human();

class Organism
{
    public Organism()
    {
        Console.WriteLine("Organism is created");
    }
}

class Human : Organism
{
    public Elephant()
    {
        Console.WriteLine("Elephant is created");
    }
}
```
In this program, we have two classes. A base Organism class and a derived Elephant class.

### b) Encapsulation:

Encapsulation is a process of binding the data members and member functions into a single unit.
Example for encapsulation is class. A class can contain data structures and methods.
Consider the following class

```csharp
public class Rectangle
{
public Rectangle ()
{
}
protected double height;
protected double width;
protected double thickness;
public double get volume()
{
Double volume=height * width * thickness;
if (volume<0)
return 0;
return volume;
}
}
``` 


### c) Abstraction:

Abstraction is a process of hiding the implementation details and displaying the essential features.
Abstraction and encapsulation are related features in object-oriented programming. Abstraction allows making relevant information visible and encapsulation enables a programmer to implement the desired level of abstraction.
Abstraction can be achieved using abstract classes in C#. 

```csharp
abstract class class_name
{
//abstract method
public abstract <return_type> method_name();
//non-abstract method
public <return_type> method_name()
{
//user code
}
}
``` 

### d) Polymorphism 
 Polymorphism means "many forms", and it occurs when we have many classes that are related to each other by inheritance.
 
```csharp
class Baby // Base class (parent) 
{
  public void babyPlay() 
  {
    Console.WriteLine("The child is playing");
  }
}

class HideSeek : Baby  // Derived class (child) 
{
  public void babyPlay() 
  {
    Console.WriteLine("The child is playing hide and seek");
  }
}

class Cooking : Baby  // Derived class (child) 
{
  public void babyPlay() 
  {
    Console.WriteLine("The child is cooking");
  }
}
```

### _Concept of OOP and Functional Programming_
 - Object-oriented programming is a software programming model based on the concept of objects, instead of just functions and procedures. As the name suggests, OOP uses objects in programming which are organized into classes, allowing individual objects to be grouped together.
 -  Functional programming, on the other hand, is a programming methodology that emphasizes on the use of function calls as the primary programming construct. It is a style of programming where you focus on the evaluation of expressions rather than the execution of commands. It provides practical approaches to problem solving in general and insights into many aspects of computing.

**Comment by sarath534** [comment, marked as answer]
# OOPs Concept:
Object-oriented programming (OOP) is a programming paradigm that uses objects, which are instances of classes, to represent and manipulate data. Classes define the properties and methods that objects of that class have.
# Functional Programming:
Functional programming, on the other hand, is a programming paradigm that emphasizes the use of functions to perform computations. Functions in functional programming are typically pure, meaning that they do not have side effects and always return the same output for a given input.
## Comparison between OOP and Functional programming
Both OOP and functional programming have their own strengths and weaknesses, and many modern programming languages support elements of both paradigms. OOP is often used for modeling real-world objects and systems, while functional programming is often used for performing complex computations and data manipulation.
# Principals of OOPs
- ## Inheritance :
 This principle states that a class can inherit the properties and methods of another class.
```csharp
class Father {
   public void Display() {
      Console.WriteLine("Display");
   }
}
class Son : Father {
   public void DisplayOne() {
      Console.WriteLine("DisplayOne");
   }
}
```
 ## Encapsulation:
- It is a technique to implement abstraction in code. Create classes and their members with appropriate access modifiers to show or hide details and complexity.

- Encapsulation hides the data and implementation details show only the required members within a class, thus hiding complexity from other code. No other code needs to know about implementation detail and also can’t modify the code of the class’s data and methods.
```csharp
class BankAccount
{
    private decimal balance;

    public void Deposit(decimal amount)
    {
        balance += amount;
    }

    public void Withdraw(decimal amount)
    {
        balance -= amount;
    }

    public decimal GetBalance()
    {
        return balance;
    }
}
```
## Polymorphism: 
Polymorphism is a Greek word that means multiple forms or shapes. You can use polymorphism if you want to have multiple forms of one or more methods of a class with the same name.

In C#, polymorphism can be achieved in two ways:
- Compile-time Polymorphism
- Run-time Polymorphism
### Example:
```csharp
class Animal  // Base class (parent) 
{
  public void animalSound() 
  {
    Console.WriteLine("The animal makes a sound");
  }
}

class Pig : Animal  // Derived class (child) 
{
  public void animalSound() 
  {
    Console.WriteLine("The pig says: wee wee");
  }
}

class Dog : Animal  // Derived class (child) 
{
  public void animalSound() 
  {
    Console.WriteLine("The dog says: bow wow");
  }
}
```
## Abstraction:
Data abstraction is the process of hiding certain details and showing only essential information to the user.
Abstraction can be achieved with either abstract classes or interface.

- Abstract class: is a restricted class that cannot be used to create objects. To access it, it must be inherited from another class.
- Abstract method: can only be used in an abstract class, and it does not have a body. The body is provided by the derived class.
### Example:
```csharp
abstract class animal {    //abstract class
    public abstract void eat();    //abstract method
    public void sound() {    
        Console.WriteLine("dog can sound");    
    }    
}    
class dog: animal {    
    public override void eat() {    
        Console.WriteLine("dog can eat");    
    }    
} 
```

**Comment by Jiby-Joseph-123** [comment]
# Object Oriented Programming.
- Object-oriented programming gives importance to data rather than procedure.
- It is based on the object-oriented concept.
- Classes are used where instances of objects are created.
- Has three access specifiers namely, Public, Private and Protected
- Provides data hiding. Hence, secured programs are possible.
```csharp
class Animal
{
    public string Name; // variables
    public string Color;
    public Animal(string name, string color)
    {
        this.Name = name;
        this.Color = color;
    }
    public void display() // method
    {
        Console.WriteLine("Iam an animal");
    }
}
class main
{
    public static void Main()
    {
        var obj = new Animal("Puppy", "Black"); // object 
        obj.display();
        
    }
}
```
In this program Animal and main are classes,  display() is a method and obj is an object created for the class Animal. 

 # abstraction:

Abstraction allows you to create seamless programs by just knowing what method to call and what parameters to input. 

example ;
```csharp 
abstract class Animal // abstract class.
{
    public abstract void animalSound(); // abstract method.

    public void sleep() //regular method.
    {
        Console.Write("Zzz");
    }
}

class Pig : Animal
{
    public override void animalSound()
    {
        Console.Write("Sound of pig!");
    }
}
class MainClass
{
    static void Main()
    {
        Pig pig = new Pig();
        pig.animalSound();
        pig.sleep();
    }
}
```
# Encapsulation:

It is a group of properties and members under a single class or object. We can use this principle to encapsulate a set of objects into different classes, thus we can prevent the repetition of code and shorten the length of the code.

# Inheritance:

Inheritance is the ability to acquire the properties of existing classes and create new ones. It allows the reuse of code without having to rewrite it in a program. 
The inheritance concept is based on a base class and a derived class.
Base class - is the class from which features are to be inherited into another class.
Derived class - it is the class in which the base class features are inherited.

Example; 
```csharp
// base class
class Animal
{
    public string Name;
    public string Color;
     public Animal(string name, string color)
    {
        this.Name = name;
        this.Color = color;
    }
    public void display()
    {
        Console.WriteLine("Iam an animal");
    }
}
// derived class

class Dog : Animal
{
    public int Age;
    public Dog(string name, string color,int age):base(name,color)
    {
    this.Age = age;
    }
    public void display()
    {
        Console.WriteLine($"Iam  {Name}, my color is {Color} age {Age}");
    }
}

class main
{
    public static void Main()
    {
        var obj = new Animal("Puppy", "Black");
        obj.display();
        var obj1 = new Dog("Pupp", "Brown",2);
        obj1.display();
    }
}
```
It is an example of single inheritance. Here we reuse the properties of Animal(base class) in  Dog(derived class) class.

# Polymorphism:

Polymorphism means many forms. It is the ability of one function to perform in different ways.

Two different ways to apply polymorphism,

- Method Overloading:
When a class has multiple methods with the same names but different sets of parameters.
 ```csharp
using System;

class Overloading
{
    void display(int a)
    {
        Console.WriteLine("Single argument:" + a);
    }
    void display(int a, int b)
    {
        Console.WriteLine($"Arguments are {a} and {b}");
    }
    static void Main()
    {
        Overloading obj = new Overloading();
        obj.display(22);
        obj.display(22, 33);
    }
}

```
- Method Overriding

# Functional programming.

- Importance is given to functions.
- It emphasizes the use of functions where each function performs a specific task.
- Data in the function are immutable.
- Fundamental elements used are variables and functions.
- Does not have any access specifier.

**Comment by ajaljosep** [comment]
Object-oriented programming (OOP) is a computer programming model that organizes software design around data, or objects

**Comment by SonuSanilK** [comment]
**Object Oriented Programming**

Object-oriented programming (OOP) is a computer programming model that organizes software design around data, or objects, rather than functions and logic. An object can be defined as a data field that has unique attributes and behavior.
OOP focuses on the objects that developers want to manipulate rather than the logic required to manipulate them. This approach to programming is well-suited for programs that are large, complex and actively updated or maintained. This includes programs for manufacturing and design, as well as mobile applications; for example, OOP can be used for manufacturing system simulation software.
The organization of an object-oriented program also makes the method beneficial to collaborative development, where projects are divided into groups. Additional benefits of OOP include code reusability, scalability and efficiency.

The object-oriented programming include the following:

- Classes
- Objects
- Methods
- Attributes

Object-oriented programming is based on the following principles:
- Encapsulation- This principle states that all important information is contained inside an object and only select information is exposed.
- Abstraction -  Objects only reveal internal mechanisms that are relevant for the use of other objects, hiding any unnecessary implementation code.
- Inheritance - Classes can reuse code from other classes. Relationships and subclasses between objects can be assigned, enabling developers to reuse common logic while still maintaining a unique hierarchy. 
- Polymorphism - Objects are designed to share behaviors and they can take on more than one form. The program will determine which meaning or usage is necessary for each execution of that object from a parent class, reducing the need to duplicate code.

**Comment by AnitaMaryThomas** [comment]
**#OOPs Concepts**
- Object Oriented Programming is a methodology to design a program using classes and objects. It follows bottom-up approach.

**#Functional Programming**

- Functional Programming paradigm emphazises on the use of functions where each function performs a specific task.
- Functional Programming don't support flow controls and conditional statements. It uses immutable data.


**#Principles of OOPs**

- The Object Oriented Programming mainly have four principles. They are: 
**1. Abstraction**
	- Abstraction is hiding the implementation details and showing only functionality to the user.
	- The abstraction can be achieved through abstract classes or interface.

**Abstract Class**

A class is declared as abstract is known as an abstract class. 
It can have abstract and non-abstract class.

**Interface**

- Interface can have only abstract methods. Interface uses interface keyword to declare an interface.

**Example**:
``` csharp
	abstract class Animal
{
public abstract void animalSound();
public void sleep()
{
Console.WriteLine("Hello");
}
}
class Lion : Animal
{
public override void animalSound()
{
Console.WriteLine("Sound of Lion");
}
}
class Mainclass
{
static void  Main()
{
Lion lion = new Lion9);
lion.animalSound();
lion.sleep();
}
}
```

**2. Inheritance**

- Inheritance is a mechanism in which one object acquires all the properties and behaviors of a parent object. 
- It is an important part of OOPs.
- It is based on a base class and a derived class.

**#Example:** 

```csharp
class Parent
{
public void Display()
{
Console.WriteLine("Hello World");
}
}
class Child : Parent
{
public void Print()
{
Console.WriteLine("Hai");
}
}
```
**3. Encapsulation**

Encapsulation is a process of wrapping code and data together into a single unit and also hiding complexity from other code.

**4. Polymorphism**

- Polymorphism means that 'many forms', and it occurs when we have many classes that are related to each other by inheritance. 

**Example:** 
```csharp
using System;
class Rectangle
{
void Display(int l)
{
Console.WriteLine("Argument is "+ l);
}
void Display(int l, int b)
{
Console.WriteLine($"Arguments are {l} and {b}");
}
static void Main()
{
Rectangle rect = new Rectangle();
rect.Display(10);
rect.Display(10, 20);
}
}
```

**Comment by SreelakshmiSajeev** [comment]
#OOP : Object Oriented Programming

In object oriented programming the given problem will be broken into objects and emphasis the object. The main aim of  OOP is to bind both class and object into single unit called class. The OOPs that include the concept such as:

-1.Object : The real time entity is called as object.

-2. Class : The collection of objects are called as class.

#Example : Class and Object

```C#
using System;

namespace MyApplication
{
  class Animal
  {
    string color = "brown";

    static void Main(string[] args)
    {
      Animal obj = new Animal();
      Console.WriteLine(obj.color);
    }
  }
}
-2.Encapsulation : The process of wrapping up of data and functions into single unit called   class is termed as encapsulation.
 
#Example.
```C#
using System;

namespace MyApplication
{
  class Animal
  {
    string color = "brown";
    public void Dispaly()
    {
       Console.WriteLine("Iam a puppy");
       
    }
    static void Main(string[] args)
    {
      Animal obj = new Animal();
      Console.WriteLine(obj.color);
      obj.Display();
    }
  }
}

-3.Abstraction 
 #Example
```C#
using System;

namespace MyApplication
{
  abstract class Animal
  {
    public abstract void animalSound();

  }
  class Dog : Animal
  {
    public override void animalSound()
    {
      Console.WriteLine("bark bark");
    }
  }

  class Program
  {
    static void Main(string[] args)
    {
      Dog obj = new Dog();  
      obj.animalSound();
     
    }
  }
}

-4.Polymorphism : The ability to show more than one form.
#Example
```C#

using System;

namespace MyApplication
{
  class Animal  
  {
    public virtual void animalDance()
    {
      Console.WriteLine("The animal can a dance");
    }
  }

  class Dog : Animal  
  {
    public override void animalDance()
    {
      Console.WriteLine("Hai I can dance");
    }
  }

 
  class Program
  {
    static void Main(string[] args)
    {
      Animal obj = new Animal();  
      Animal dog = new Dog();  
      
      myAnimal.animalDance();
      dog.animalDance();
    }
  }
}

-5.Inheritance : The process of inheriting the properties of another class .

#Example
```C#

using System;

namespace MyApplication
{
  class Animal  
  {
    public virtual void animalDance()
    {
      Console.WriteLine("The animal can a dance");
    }
  }

  class Dog : Animal  
  {
    public override void animalDance()
    {
      Console.WriteLine("Hai I can dance");
    }
  }

 
  class Program
  {
    static void Main(string[] args)
    {
      Animal obj = new Animal();  
      Animal dog = new Dog();  
      
      myAnimal.animalDance();
      dog.animalDance();
    }
  }
}

**Comment by kavitha-BS** [comment]
[oops.odt](https://github.com/mishelshaji/Learn-DotNet/files/10530520/oops.odt)
